wren-ui/src/apollo/server/managers/dataSourceSchemaDetector.ts

```diff
@@ -0,0 +1,195 @@
+import { differenceWith, isEmpty, isEqual } from 'lodash';
+import { CompactTable } from '@server/connectors/connector';
+import { DataSourceStrategyFactory } from '@server/factories/onboardingFactory';
+import { IContext } from '@server/types';
+import { getLogger } from 'log4js';
+
+const logger = getLogger('DataSourceSchemaDetector');
+logger.level = 'debug';
+
+type DataSourceSchema = {
+  name: string;
+  columns: {
+    name: string;
+    type: string;
+  }[];
+};
+
+type DataSourceSchemaChange = {
+  [SchemaChangeType.DELETED_TABLES]: DataSourceSchema[];
+  [SchemaChangeType.DELETED_COLUMNS]: DataSourceSchema[];
+  [SchemaChangeType.MODIFIED_COLUMNS]: DataSourceSchema[];
+};
+
+enum SchemaChangeType {
+  // the tables has been deleted
+  DELETED_TABLES = 'deletedTables',
+  // the columns has been deleted
+  DELETED_COLUMNS = 'deletedColumns',
+  // the columns type has been changed
+  MODIFIED_COLUMNS = 'modifiedColumns',
+}
+
+interface IDataSourceSchemaDetector {
+  detectSchemaChange(): Promise<void>;
+}
+
+export default class DataSourceSchemaDetector
+  implements IDataSourceSchemaDetector
+{
+  public ctx: IContext;
+  public projectId: string;
+
+  constructor({ ctx }: { ctx: IContext }) {
+    this.ctx = ctx;
+  }
+
+  public async detectSchemaChange() {
+    const diffSchema = await this.getDiffSchema();
```

Comment:
We can pass the project from the resolver. It will be easier to adapt to the multiple-project scenario